issue/6111-step4-site-picker-changes

anitaa1990 opened the pull request:
Part 4 of #6111. This PR is the final step of the site picker refactor. 

## Changes
- Added separate views for empty view (if there are no sites found) and user info view. We previously had logic to show or hide the user account details in the site picker screen. But following iOS here, I updated the UI to reflect iOS.
- Updated the fragment UI + skeleton views for the site picker.
- Updated WooUpgradeRequiredDialog and SitePickerAdapter to latest code updates.
- Removed redundant classes like the `SitePickerActivity`, `SitePickerContract` and `SitePickerPresenter`.

## Testing instructions
#### Switch Stores
- Click on the Hub menu icon -> Switch store button.
- Notice that the list of sites is displayed along with the user account info and the continue button.
- Click on one of the sites in the list and notice the progress dialog is displayed and the My Store tab is refreshed to reflect the currently selected site.

#### Login via site address: auto login to the correct site address
- Click on the `Login with site address` button. Enter a store url, email + password.
- Notice you are redirected to the site picker screen. 
- Progress dialog is displayed with `Verifying site` and you are redirected to the My Store tab for the entered site address.

#### Login via site address: auto login to a site address entered during login.
- Click on the `Login with site address` button. Enter a store url, email + password.
- Notice you are redirected to the site picker screen. 
- The current site is not found in the list of stores associated with the logged in user account so account mismatch view is displayed. 
- **If there are sites associated with the user account**, then `View connected stores` button is displayed.
- Click on `View Connected Stores` would redirect to a list of stores associated with the account.
- Click on `Try Another account` will redirect to the prologue screen after logging out the user.
- **If there are no sites associated with the user account**, then an empty view is displayed with the option to login with a different account and an option to install Jetpack on their store.
- **If the entered site address does not have WooCommerce installed**, then install Woo option is displayed to the merchant. They can install Woo and refresh the screen, if needed. 

#### Login via email to an account that does not have any stores:
- Click on the `Login with WordPress.com` button. Enter email + password.
- Notice you are redirected to the site picker screen. 
- Notice that if there are no sites associated with this account, then an empty view is displayed with the option to login with a different account and an option to install Jetpack on their store.
- Click on `Try Another account` will redirect to the prologue screen after logging out the user.

#### Jetpack CP flow
- The above testing steps should work for Jetpack CP sites too. It would be nice to test those scenarios using a Jetpack CP site.

## Images/gif
#### Switch Store flow
<img src="https://user-images.githubusercontent.com/22608780/164007942-5fbcfff0-d74e-4167-a7bd-6529ae2e9d4f.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/164007952-d85ecff4-1f84-4fbc-9eec-fafa7d5f8791.png" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/164008272-44417cde-cb30-41ed-b88d-dd0972a2053f.gif" width="300"/>

#### Login via store address flow
<img src="https://user-images.githubusercontent.com/22608780/164008457-7db5e2be-d8b5-4785-877a-ef7036840923.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/164008468-57aa4192-9238-46a9-8c72-619aa2f34779.gif" width="300"/>

#### Login via email flow
<img src="https://user-images.githubusercontent.com/22608780/164008626-574be9b8-0f21-401c-8fb2-4c26647aa126.gif" width="300"/>

#### Login via email flow - No sites found
<img src="https://user-images.githubusercontent.com/22608780/164008639-7780f8a7-ac71-42d4-afe9-a58914227cb7.png" width="300"/>

#### Auto login to a site if logged in via site address flow - The site address entered is not associated with the user account
<img src="https://user-images.githubusercontent.com/22608780/164008889-91d28154-dac9-4ef6-873d-0e3ac0db273c.png" width="300"/>

#### Auto login to a site if logged in via site address flow - Woo not found
<img src="https://user-images.githubusercontent.com/22608780/164008939-d033232b-478c-4773-9d05-63e43f3bb313.png" width="300"/>

#### Auto login to a site if logged in via site address flow - Woo not found
<img src="https://user-images.githubusercontent.com/22608780/164008997-1bf79468-7b1f-42c0-9c77-05784b9b8385.png" width="300"/>

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

anitaa1990 committed:
Added separate views for empty site picker view + user info view

anitaa1990 committed:
Updated WooUpgradeRequiredDialog and SitePickerAdapter

anitaa1990 committed:
Updated sitepicker fragment UI + skeleton views

anitaa1990 committed:
Removed redundant second call to open site picker when app is first opened

anitaa1990 committed:
Added fixes for account mismatch view

anitaa1990 committed:
Updated UI for no woo store found

anitaa1990 committed:
Updated unit tests

anitaa1990 committed:
Fixed lint issues

anitaa1990 committed:
Removed redundant classes

anitaa1990 committed:
Merge branch 'issue/6111-step3-site-picker-changes' into issue/6111-step4-site-picker-changes

anitaa1990 committed:
Fixed lint issues

anitaa1990 committed:
Fixed unit tests build failure

anitaa1990 committed:
Fixed detekt issues once again in unit tests

JorgeMucientes commented on WooCommerce/src/main/AndroidManifest.xml:
```diff
@@ -45,10 +45,6 @@
         </activity>
 
-        <activity
-            android:name=".ui.sitepicker.SitePickerActivity"
```
Np, now that `SitePickerActivity` has been deleted I noticed that there are some "leftovers" for Detekt `baseline.xml` file that can be removed.

anitaa1990 replied:
Done in [35f07c8](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/35f07c81cebc0b4fb5dc0b9ba04448616262620a)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt:
```diff
@@ -170,13 +170,6 @@ class AppInitializer @Inject constructor() : ApplicationLifecycleListener {
             dispatcher.dispatch(AccountActionBuilder.newFetchSettingsAction())
             appCoroutineScope.launch {
                 wooCommerceStore.fetchWooCommerceSites()
-                if (!selectedSite.exists() && ProcessLifecycleOwner.get().lifecycle.currentState.isAtLeast(STARTED)) {
-                    // The previously selected site is not connected anymore, take the user to the site picker
-                    WooLog.i(DASHBOARD, "Selected site no longer exists, showing site picker")
-                    val intent = Intent(application, MainActivity::class.java)
-                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
-                    application.startActivity(intent)
-                }
```
@anitaa1990 I don't think removing this logic is safe, this was added to account for a special case:
1. The previously selected site in the app gets deleted from the user's account while the app was being closed.
2. The user opens the app, the app will fetch the user's sites asynchronisly.
3. In MainActivity, we won't know that the site has been deleted, as the call is asynchronous, so we will let the user use the app.
4. When the API call finishes, the site will be removed from the DB, and this would lead to the [crash](https://github.com/woocommerce/woocommerce-android/issues/4842).

I saw your comment on the other [PR](https://github.com/woocommerce/woocommerce-android/pull/6295#issuecomment-1118363536), so I know why this logic has been removed, but IMO it should be improved instead of removing it completely, if we want to avoid restarting the whole activity, we can update the logic to make sure we call `selectedSite.reset()` after the fetching finishes, then update `MainActivity` to observe `SelectedSite` and route to the Site Picker when the `Flow` returns a `null` value.

hichamboushaba replied:
A alternative and simpler way to fix this would be to update the check in the `if` to see if the user has a selected site previously, so something like this:

```kotlin
    if (selectedSite.getSelectedSiteId() != -1 &&
        !selectedSite.exists() &&
        ProcessLifecycleOwner.get().lifecycle.currentState.isAtLeast(STARTED)
    ) {
        // The previously selected site is not connected anymore, take the user to the site picker
        WooLog.i(DASHBOARD, "Selected site no longer exists, showing site picker")
        val intent = Intent(application, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        application.startActivity(intent)
    }
```
This would solve the edge case that Jorge identified and which happens only if the user didn't select a site previously.

anitaa1990 replied:
Thanks for the context here @hichamboushaba ❤️  ! I honestly didn't realise this was handling a special case and just thought since we changed the logic in the `MainActivity`, we could just remove this logic 🤦‍♀️ . I added your suggested solution in [48c3234](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/48c323472a4cc1132266377ceca83c1265beff74).

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerAdapter.kt:
```diff
@@ -1,103 +1,76 @@
-    }
-
+class SitePickerAdapter(
+    private val onSiteSelected: (SiteModel) -> Unit
```
Np, I think you can add this function inside: 
```kotlin
    @Parcelize
    data class SiteUiModel(
        val site: SiteModel,
        val isSelected: Boolean,
        val onSiteSelected: (SiteModel) -> Unit
    ) : Parcelable
```

And then get rid of this constructor dependency as well as this line in `SitePickerFragment`
```kotlin
    private val sitesAdapter by lazy { SitePickerAdapter(viewModel::onSiteSelected) }
```

anitaa1990 replied:
That's a good approach! I tried to refactor it and there were a lot of places that needed a change so I decided to leave it as is for now? Is that ok?

hichamboushaba replied:
I think this is OK.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerAdapter.kt:
```diff
@@ -1,103 +1,76 @@
         setHasStableIds(true)
     }
 
-    override fun getItemId(position: Int): Long = getItem(position).siteId
```
hmm, I think changing this adapter to `RecyclerView.Adapter<SiteViewHolder>()` instead of `ListAdapter` and then removing this `fun getItemId()` override will have unwanted behaviors in the UI. Basically, when clicking on any site and updating the list to mark the site as selected, the list will be reordered because items don't have a unique id anymore. This is the effect:

https://user-images.githubusercontent.com/2663464/166971658-1d062bda-7045-446b-aed8-4fa639dee20c.mp4

Out of curiosity, what made you switch to use `RecyclerView.Adapter` instead of `ListAdapter` with `SiteModelDiffCallBack`

anitaa1990 replied:
Awesome catch @JorgeMucientes! I think I just used what we use in a lot of places around the app 😅 I ended up adding a `getItemId` method to the `SitePickerAdapter` instead of converting it into a `ListAdapter`. I didn't think there was a significant improvement by converting it to a `ListAdapter` but I could be wrong here and totally out of context 😄

hichamboushaba replied:
This looks good, I don't think `ListAdapter` would bring a significant improvement for this case, as generally we will have a small number of sites, and a small number of changes.

JorgeMucientes replied:
Just to provide a bit more context, but no need to change this. 
I have nothing against `RecyclerView.Adapter<SiteViewHolder>` approach. The thing is that there was [some work done](https://github.com/woocommerce/woocommerce-android/pull/6174) recently,  in an attempt to migrate the adapters of the app to keep a consistent approach on how to apply `Diffs` over items in `recyclerviews`. Although woudl have to double check with @nbradbury if that was the case here.

 I also think that the `ListAdapter` approach has 2 advantages: 
- Less code needed
- `submitList` handles diffing in a background thread instead of doing the diff in the main thread. Which in this case as Hicham mentioned is irrelevant because the list of sites most likely is going to be super short. 

That said, no need to change anything 🙂. I was simply curious about why "reverting" this adapter to `RecyclerView.Adapter`approach after the last changes.

nbradbury replied:
It would be nice to keep the `ListAdapter` approach I worked on recently, unless there's a reason it had to be changed.

anitaa1990 replied:
No specific reason other than I took code reference from Order detail adapters that seem to be still using `RecyclerView.Adapter`. I'll update it to `ListAdapter` if that is what we are primarily using in the app. Thanks for the explanation and context @JorgeMucientes @nbradbury

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/views/LoginUserInfoView.kt:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    fun leftAlign() {
+        binding.loginUserInfo.gravity = Gravity.START
+        with(binding.imageAvatar) {
+            layoutParams.height = resources.getDimensionPixelSize(dimen.image_major_72)
+            layoutParams.width = resources.getDimensionPixelSize(dimen.image_major_72)
+            requestLayout()
+        }
+    }
```
Np, can we remove this now that we are center aligning the user details?

anitaa1990 replied:
👍 Fixed in [d9777d4](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/d9777d4d69b28a254d2a8fe3db9153b5b91c21f7)

JorgeMucientes commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModelTest.kt:
```diff
@@ -523,16 +524,28 @@ class SitePickerViewModelTest : BaseUnitTest() {
 
     @Test
-    fun `Given user is logging in, when view connected sites is clicked, screen is refreshed`() = testBlocking {
+    fun `Given user is logging in, when view connected sites is clicked, site list is displayed`() = testBlocking {
```
Super minor np, (in fact I'm guilty of doing this all the time 😅 ) but "Given" word should be lower case "given" afaik

anitaa1990 replied:
Fixed in [944dd84](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/944dd848d5973a73f1cd40558be278fd39d36e0f)

JorgeMucientes commented:
Great job @anitaa1990 🙌  🏅  Tested all the cases and works as expected 👍🏼 

 I just left some minor nitpicks and one UI issue regarding the site adapter items moving around. Feel free to apply or not the rest of the nitpicks 🙂

anitaa1990 committed:
Changed Given -> given for unit tests in SitePickerViewModelTest

anitaa1990 committed:
Removed unused method in LoginUserInfoView

anitaa1990 committed:
Fixed issue with UX when selecting a site

anitaa1990 committed:
Removed redundant baseline rules

anitaa1990 committed:
Added back the logic to redirect user to MainActivity if selected site is not available

anitaa1990 commented:
Thanks for the review @JorgeMucientes and @hichamboushaba. Sorry for the delay in getting back to this PR! I have addressed all the comments in the PR and it's ready for another round 👍

anitaa1990 commented:
cc @hichamboushaba @0nko , since Jorge is on his maintenance rotation this week and next, would one of you mind taking over the review of this PR please 🙏  ?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerFragment.kt:
```diff
@@ -1,8 +1,249 @@
+
+    private val viewModel: SitePickerViewModel by viewModels()
+
+    private val sitesAdapter by lazy { SitePickerAdapter(viewModel::onSiteSelected) }
```
Just sharing something that I learned recently, this usage will cause a memory leak, I was guilty of using too in the past, until we [identified](https://github.com/woocommerce/woocommerce-android/pull/5512#issuecomment-1000406011) a memory leak in OrderCreationFormFragment, and we identified this usage as the cause.

The fix was simple: https://github.com/woocommerce/woocommerce-android/pull/5512/commits/72a1383f1ceeb20046fb0571f670ad2d04f8ce06

anitaa1990 replied:
Thanks for the context! Fixed in [7ce7269](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/7ce7269f409612a9cfc15e1def348df7a1fe4150)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerFragment.kt:
```diff
@@ -1,8 +1,249 @@
+                else -> event.isHandled = false
+            }
+        }
+        viewModel.start()
```
As shared on a previous PR, having an explicit `start` function breaks the lifecycle consideration for the Fragment, as we will be re-starting the ViewModel on each view re-creation, and this happens on multiple events: screen rotation, navigation...

For this specific screen, this introduces two major issues:
1. We lose the selected site on each screen rotation.
2. We re-fetch the list of sites on each screen rotation.

The second point is even worse when the list of sites has a Jetpack CP site, as we skip the DB sites to make sure the workarounds we have for Jetpack CP sites is executed, check this recording*:

https://user-images.githubusercontent.com/1657201/167903704-7d13039d-e6b3-4abf-9e4d-dc7a471792bb.mp4 

I think the ViewModel initialization needs to be moved to the `init` block, to make sure it survives configuration changes, this won't solve the first point when "don't keep activities" is enabled, but this is an edge case, and it shouldn't be a blocking point.

\* to reproduce the Jetpack CP issue, please make sure you are using a WCPay version before 4.0.0, as it introduced a bug that makes the API to return `"jetpack":true`, which makes the site looks like it has full jetpack. I will report this bug to the wcpay team.

anitaa1990 replied:
This is a good catch! As I had mentioned in that PR, I was using the init function but moved to using start due to some issues with testing. I changed it to back now in [5f3d2ab](https://github.com/woocommerce/woocommerce-android/pull/6298/commits/5f3d2ab314f0ec35da4a7bff12c666fe7cb1f875)

hichamboushaba reviewed (commented):
Nice work @anitaa1990, I didn't test all cases, as I assume @JorgeMucientes test them all.

But I found some issues that are caused by the usage of the `start` function, please review my comments, then let me know how you want to proceed, if you want to address this in a separate PR, I'm not against it.

And a np about something that I missed in my review for the PR #6288, I think we shouldn't use an `Event` class for the Ui state, I mean here the sealed class `SitePickerView`, IMO this should be just a simple Enum class that's part of the `ViewState` class.
I think this is not causing an issue now, but when we move away from using the `start` function, it would show why having it as part of the `ViewState` is a better approach, as using an event might cause losing the state after screen rotation, while with the `ViewState` we will keep it, and the Fragment will use the correct state always when it starts observing it.

anitaa1990 committed:
Updated SitePickerAdapter to use ListAdapter

anitaa1990 committed:
Removed start method from SitePickerViewModel

anitaa1990 commented:
Thank you for the review @hichamboushaba! I fixed the two issues mentioned in the comment. Regarding this comment:

> And a np about something that I missed in my review for the PR https://github.com/woocommerce/woocommerce-android/pull/6288, I think we shouldn't use an Event class for the Ui state, I mean here the sealed class SitePickerView, IMO this should be just a simple Enum class that's part of the ViewState class.
I think this is not causing an issue now, but when we move away from using the start function, it would show why having it as part of the ViewState is a better approach, as using an event might cause losing the state after screen rotation, while with the ViewState we will keep it, and the Fragment will use the correct state always when it starts observing it.

Would it be ok to handle this in a separate PR? This one is already too big and since it's getting merged to a feature branch, I can work on it separately before merging that feature branch to trunk

hichamboushaba reviewed (approved):
Thanks @anitaa1990 for making those changes, let's :shipit:

> Would it be ok to handle this in a separate PR? This one is already too big and since it's getting merged to a feature branch, I can work on it separately before merging that feature branch to trunk

Yes, this is OK.
Please ping me later if my explanation of the issue above is not clear, the fix should be pretty simple, just adding a field to the ViewState class to hold the `SitePickerView` value, and set it instead of triggering it as event.
And just to confirm, as I suspected, removing the `start` would break the logic now, the `buttonPrimary` listener doesn't work after screen rotation.
